src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
         if(!initialized)
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```

Comment:
Yeah, I get that `WrappedAuthenticationProvider` is essentially transforming an `AuthenticationProvider` implementation into a `ServerAuthenticationProvider`. My concern is that it feels like the wrong abstraction to me because the implementations we currently have are independent of `ZooKeeperServer` instance. 

The fact that `ServerAuthenticationProvider` depends on a `ZooKeeperServer` instance is specific to the class of providers the interface is representing, but not all. The abstraction doesn't quite work in my head when I think that we are trying to make all `AuthenticationProvider` implementations an instance of `ServerAuthenticationProvider` via the wrapper.

In general, I get how it works, I'm just not super convinced that we are reflecting the right abstraction in the code.